In [2]:
import math
import numpy as np
import pandas as pd

We read the dataset provided by using the pandas  library as follows:

In [3]:
df=pd.read_csv("Worksheet in Case Discussion_Grocesory Store Demand Analysis.csv")
df.head()

Date  Milk_Sales (Gallons)  Milk_Forecast (Gallons)  \
0  1/1/2024                   200                      205   
1  1/2/2024                   195                      190   
2  1/3/2024                   198                      200   
3  1/4/2024                   202                      200   
4  1/5/2024                   199                      200   

   Bread_Sales (Loaves)  Bread_Forecast (Loaves)  Bananas_Sales (Pounds)  \
0                   150                      155                     300   
1                   155                      150                     310   
2                   148                      150                     295   
3                   160                      160                     280   
4                   152                      150                     290   

   Bananas_Forecast (Pounds)  Chicken_Sales (Pounds)  \
0                        310                     180   
1                        300                     175   
2                        300                     190   
3                        275                     185   
4                        295                     170   

   Chicken_Forecast (Pounds)  Toilet_Paper_Sales (Packs of 12 Rolls)  \
0                        175                                     100   
1                        170                                     110   
2                        185                                     120   
3                        190                                     130   
4                        165                                     105   

   Toilet_Paper_Forecast (Packs of 12 Rolls)  
0                                        105  
1                                        115  
2                                        125  
3                                        135  
4                                        100

Now, to calculate the demand patterns for eachc product, I will create dataframes solely for the five different products.

Starting with milk first, this is done as follows: 


In [4]:
milk=pd.DataFrame()
milk=pd.concat([milk,df['Date'],df['Milk_Sales (Gallons)'],df['Milk_Forecast (Gallons)']],axis=1)
milk.head()


Date  Milk_Sales (Gallons)  Milk_Forecast (Gallons)
0  1/1/2024                   200                      205
1  1/2/2024                   195                      190
2  1/3/2024                   198                      200
3  1/4/2024                   202                      200
4  1/5/2024                   199                      200

In [5]:
milk_actual=milk['Milk_Sales (Gallons)']
milk_actual.head()

0    200
1    195
2    198
3    202
4    199
Name: Milk_Sales (Gallons), dtype: int64

In [6]:
milk_fore=milk['Milk_Forecast (Gallons)']
milk_fore.head()

0    205
1    190
2    200
3    200
4    200
Name: Milk_Forecast (Gallons), dtype: int64

Now, we start by calculating the different types of errors between the actual sales and the forecast sales

In [7]:
error=[]
abs_error=[]
perc_error=[]
sq_error=[]
for i in range(len(milk_actual)):
    ele1=milk_actual[i]-milk_fore[i]
    ele2=int(math.fabs(ele1))
    ele3=(ele2/milk_actual[i])*100
    ele4=ele1**2
    error.append(ele1)
    abs_error.append(ele2)
    perc_error.append(ele3)
    sq_error.append(ele4)

print(error)
print(abs_error)
print(perc_error)
print(sq_error)

[-5, 5, -2, 2, -1, 0, -5, -5, 5, -2, -3, 0, 0, -2, 2, 4, 0, -5, -5, 5, -2, -3, 0, 0, -2, 2, 4, 0, -5, -5]
[5, 5, 2, 2, 1, 0, 5, 5, 5, 2, 3, 0, 0, 2, 2, 4, 0, 5, 5, 5, 2, 3, 0, 0, 2, 2, 4, 0, 5, 5]
[2.5, 2.564102564102564, 1.0101010101010102, 0.9900990099009901, 0.5025125628140703, 0.0, 2.3255813953488373, 2.272727272727273, 2.4390243902439024, 1.0101010101010102, 1.4851485148514851, 0.0, 0.0, 1.0101010101010102, 0.9900990099009901, 2.0100502512562812, 0.0, 2.3255813953488373, 2.272727272727273, 2.4390243902439024, 1.0101010101010102, 1.4851485148514851, 0.0, 0.0, 1.0101010101010102, 0.9900990099009901, 2.0100502512562812, 0.0, 2.3255813953488373, 2.272727272727273]
[25, 25, 4, 4, 1, 0, 25, 25, 25, 4, 9, 0, 0, 4, 4, 16, 0, 25, 25, 25, 4, 9, 0, 0, 4, 4, 16, 0, 25, 25]


Once we calculate all the errors, it is time to add them in our dataframe as follows:

In [8]:
df1=pd.DataFrame(error)
df2=pd.DataFrame(abs_error)
df3=pd.DataFrame(perc_error)
df4=pd.DataFrame(sq_error)


In [9]:
milk=pd.concat([milk,df1,df2,df3,df4,],axis=1)

In [10]:
milk.columns=['Date','Milk_Sales (Gallons)','Milk_Forecast (Gallons)','Error','Absolute Error','Percentage Error','Squared Error' ]

In [11]:
milk.head()

Date  Milk_Sales (Gallons)  Milk_Forecast (Gallons)  Error  \
0  1/1/2024                   200                      205     -5   
1  1/2/2024                   195                      190      5   
2  1/3/2024                   198                      200     -2   
3  1/4/2024                   202                      200      2   
4  1/5/2024                   199                      200     -1   

   Absolute Error  Percentage Error  Squared Error  
0               5          2.500000             25  
1               5          2.564103             25  
2               2          1.010101              4  
3               2          0.990099              4  
4               1          0.502513              1

Once they are added to the DataFrame, its time to calculate the MAPE, MAD, Forecast Bias, and Variance values.

In [12]:
mape_val=[]
for i in range(len(milk_actual)):
    ele=math.fabs((milk_actual[i]-milk_fore[i])/milk_actual[i])*100
    mape_val.append(ele)
print(mape_val)

[2.5, 2.564102564102564, 1.0101010101010102, 0.9900990099009901, 0.5025125628140703, 0.0, 2.3255813953488373, 2.272727272727273, 2.4390243902439024, 1.0101010101010102, 1.4851485148514851, 0.0, 0.0, 1.0101010101010102, 0.9900990099009901, 2.0100502512562812, 0.0, 2.3255813953488373, 2.272727272727273, 2.4390243902439024, 1.0101010101010102, 1.4851485148514851, 0.0, 0.0, 1.0101010101010102, 0.9900990099009901, 2.0100502512562812, 0.0, 2.3255813953488373, 2.272727272727273]


In [13]:
Sum=0
for i in mape_val:
    Sum+=i
mape=Sum/len(mape_val)
print(mape)

1.3083596508018773


In [14]:
MAPE=[mape]
df5=pd.DataFrame(MAPE)
milk=pd.concat([milk,df5],axis=1)

In [15]:
mad_val=[]
for i in range(len(milk_actual)):
    ele=math.fabs(milk_actual[i]-milk_fore[i])
    mad_val.append(ele)
print(mad_val)

[5.0, 5.0, 2.0, 2.0, 1.0, 0.0, 5.0, 5.0, 5.0, 2.0, 3.0, 0.0, 0.0, 2.0, 2.0, 4.0, 0.0, 5.0, 5.0, 5.0, 2.0, 3.0, 0.0, 0.0, 2.0, 2.0, 4.0, 0.0, 5.0, 5.0]


In [16]:
summ=0
for i in mad_val:
    summ+=i
mad=Sum/len(mad_val)
print(mad)

1.3083596508018773


In [17]:
MAD=[mad]
df6=pd.DataFrame(MAD)
milk=pd.concat([milk,df6],axis=1)

In [18]:
fore_val=[]
for i in range(len(milk_fore)):
    ele=milk_fore[i]-milk_actual[i]
    fore_val.append(ele)
print(fore_val)

[5, -5, 2, -2, 1, 0, 5, 5, -5, 2, 3, 0, 0, 2, -2, -4, 0, 5, 5, -5, 2, 3, 0, 0, 2, -2, -4, 0, 5, 5]


In [19]:
ssum=0
for i in fore_val:
    ssum+=i
fore=ssum/len(fore_val)
print(fore)

0.7666666666666667


In [20]:
BIAS=[fore]
df7=pd.DataFrame(BIAS)
milk=pd.concat([milk,df7],axis=1)

In [21]:
sse=milk["Squared Error"]
suum=0
for i in sse:
    suum+=i
var=suum/len(sse)
print(var)

11.1


In [22]:
VAR=[var]
df8=pd.DataFrame(VAR)
milk=pd.concat([milk,df8],axis=1)

In [23]:
milk.columns=['Date','Milk_Sales (Gallons)','Milk_Forecst (Gallons)','Error','Absolute Error','Percentage Error','Squared Error','MAPE','MAD','Forecast Bias','Variance']

In [24]:
milk.head()

Date  Milk_Sales (Gallons)  Milk_Forecst (Gallons)  Error  \
0  1/1/2024                   200                     205     -5   
1  1/2/2024                   195                     190      5   
2  1/3/2024                   198                     200     -2   
3  1/4/2024                   202                     200      2   
4  1/5/2024                   199                     200     -1   

   Absolute Error  Percentage Error  Squared Error     MAPE      MAD  \
0               5          2.500000             25  1.30836  1.30836   
1               5          2.564103             25      NaN      NaN   
2               2          1.010101              4      NaN      NaN   
3               2          0.990099              4      NaN      NaN   
4               1          0.502513              1      NaN      NaN   

   Forecast Bias  Variance  
0       0.766667      11.1  
1            NaN       NaN  
2            NaN       NaN  
3            NaN       NaN  
4            NaN       NaN

Once the metrics are calculated and added to the dataframe, we can analyze the forecast accuracy and demand variability for the respective product. 

We can see that the MAPE value for the Milk forecast is pretty low, which indicates that the forecast for Milk is off by around 1.3%, suggesting that it's an accurate forecast.

Now, we do the same for all the products.

BREAD

In [25]:
bread=pd.DataFrame()
bread=pd.concat([bread,df['Date'],df['Bread_Sales (Loaves)'],df['Bread_Forecast (Loaves)']],axis=1)
bread.head()


Date  Bread_Sales (Loaves)  Bread_Forecast (Loaves)
0  1/1/2024                   150                      155
1  1/2/2024                   155                      150
2  1/3/2024                   148                      150
3  1/4/2024                   160                      160
4  1/5/2024                   152                      150

In [26]:
bread_actual=bread['Bread_Sales (Loaves)']
bread_actual.head()

0    150
1    155
2    148
3    160
4    152
Name: Bread_Sales (Loaves), dtype: int64

In [27]:
bread_fore=bread['Bread_Forecast (Loaves)']
bread_fore.head()

0    155
1    150
2    150
3    160
4    150
Name: Bread_Forecast (Loaves), dtype: int64

In [28]:
error=[]
abs_error=[]
perc_error=[]
sq_error=[]
for i in range(len(bread_actual)):
    ele1=bread_actual[i]-bread_fore[i]
    ele2=int(math.fabs(ele1))
    ele3=(ele2/bread_actual[i])*100
    ele4=ele1**2
    error.append(ele1)
    abs_error.append(ele2)
    perc_error.append(ele3)
    sq_error.append(ele4)

print(error)
print(abs_error)
print(perc_error)
print(sq_error)

[-5, 5, -2, 0, 2, 0, -5, -5, 5, -5, -5, 0, 5, -2, 0, 0, -2, 0, 2, 0, -5, -5, 5, -5, -5, 0, 5, -2, 0, 0]
[5, 5, 2, 0, 2, 0, 5, 5, 5, 5, 5, 0, 5, 2, 0, 0, 2, 0, 2, 0, 5, 5, 5, 5, 5, 0, 5, 2, 0, 0]
[3.3333333333333335, 3.225806451612903, 1.3513513513513513, 0.0, 1.3157894736842104, 0.0, 3.3333333333333335, 3.125, 3.225806451612903, 3.3333333333333335, 3.4482758620689653, 0.0, 3.225806451612903, 1.3513513513513513, 0.0, 0.0, 1.3513513513513513, 0.0, 1.3157894736842104, 0.0, 3.3333333333333335, 3.125, 3.225806451612903, 3.3333333333333335, 3.4482758620689653, 0.0, 3.225806451612903, 1.3513513513513513, 0.0, 0.0]
[25, 25, 4, 0, 4, 0, 25, 25, 25, 25, 25, 0, 25, 4, 0, 0, 4, 0, 4, 0, 25, 25, 25, 25, 25, 0, 25, 4, 0, 0]


In [29]:
df1=pd.DataFrame(error)
df2=pd.DataFrame(abs_error)
df3=pd.DataFrame(perc_error)
df4=pd.DataFrame(sq_error)


In [30]:
bread=pd.concat([bread,df1,df2,df3,df4,],axis=1)

In [31]:
bread.columns=['Date','Milk_Sales (Gallons)','Milk_Forecast (Gallons)','Error','Absolute Error','Percentage Error','Squared Error' ]

In [32]:
bread.head()

Date  Milk_Sales (Gallons)  Milk_Forecast (Gallons)  Error  \
0  1/1/2024                   150                      155     -5   
1  1/2/2024                   155                      150      5   
2  1/3/2024                   148                      150     -2   
3  1/4/2024                   160                      160      0   
4  1/5/2024                   152                      150      2   

   Absolute Error  Percentage Error  Squared Error  
0               5          3.333333             25  
1               5          3.225806             25  
2               2          1.351351              4  
3               0          0.000000              0  
4               2          1.315789              4

In [33]:
mape_val=[]
for i in range(len(bread_actual)):
    ele=math.fabs((bread_actual[i]-bread_fore[i])/bread_actual[i])*100
    mape_val.append(ele)
print(mape_val)

[3.3333333333333335, 3.225806451612903, 1.3513513513513513, 0.0, 1.3157894736842104, 0.0, 3.3333333333333335, 3.125, 3.225806451612903, 3.3333333333333335, 3.4482758620689653, 0.0, 3.225806451612903, 1.3513513513513513, 0.0, 0.0, 1.3513513513513513, 0.0, 1.3157894736842104, 0.0, 3.3333333333333335, 3.125, 3.225806451612903, 3.3333333333333335, 3.4482758620689653, 0.0, 3.225806451612903, 1.3513513513513513, 0.0, 0.0]


In [34]:
Sum=0
for i in mape_val:
    Sum+=i
mape=Sum/len(mape_val)
print(mape)

1.7993078333880983


In [35]:
MAPE=[mape]
df5=pd.DataFrame(MAPE)
bread=pd.concat([bread,df5],axis=1)

In [36]:
mad_val=[]
for i in range(len(bread_actual)):
    ele=math.fabs(bread_actual[i]-bread_fore[i])
    mad_val.append(ele)
print(mad_val)

[5.0, 5.0, 2.0, 0.0, 2.0, 0.0, 5.0, 5.0, 5.0, 5.0, 5.0, 0.0, 5.0, 2.0, 0.0, 0.0, 2.0, 0.0, 2.0, 0.0, 5.0, 5.0, 5.0, 5.0, 5.0, 0.0, 5.0, 2.0, 0.0, 0.0]


In [37]:
summ=0
for i in mad_val:
    summ+=i
mad=Sum/len(mad_val)
print(mad)

1.7993078333880983


In [38]:
MAD=[mad]
df6=pd.DataFrame(MAD)
bread=pd.concat([bread,df6],axis=1)

In [39]:
fore_val=[]
for i in range(len(bread_fore)):
    ele=bread_fore[i]-bread_actual[i]
    fore_val.append(ele)
print(fore_val)

[5, -5, 2, 0, -2, 0, 5, 5, -5, 5, 5, 0, -5, 2, 0, 0, 2, 0, -2, 0, 5, 5, -5, 5, 5, 0, -5, 2, 0, 0]


In [40]:
ssum=0
for i in fore_val:
    ssum+=i
fore=ssum/len(fore_val)
print(fore)

0.8


In [41]:
BIAS=[fore]
df7=pd.DataFrame(BIAS)
bread=pd.concat([bread,df7],axis=1)

In [42]:
sse=bread["Squared Error"]
suum=0
for i in sse:
    suum+=i
var=suum/len(sse)
print(var)

12.466666666666667


In [43]:
VAR=[var]
df8=pd.DataFrame(VAR)
bread=pd.concat([bread,df8],axis=1)

In [44]:
bread.columns=['Date','Bread_Sales (Loaves)','Bread_Forecst (Loaves)','Error','Absolute Error','Percentage Error','Squared Error','MAPE','MAD','Forecast Bias','Variance']

In [45]:
bread.head()

Date  Bread_Sales (Loaves)  Bread_Forecst (Loaves)  Error  \
0  1/1/2024                   150                     155     -5   
1  1/2/2024                   155                     150      5   
2  1/3/2024                   148                     150     -2   
3  1/4/2024                   160                     160      0   
4  1/5/2024                   152                     150      2   

   Absolute Error  Percentage Error  Squared Error      MAPE       MAD  \
0               5          3.333333             25  1.799308  1.799308   
1               5          3.225806             25       NaN       NaN   
2               2          1.351351              4       NaN       NaN   
3               0          0.000000              0       NaN       NaN   
4               2          1.315789              4       NaN       NaN   

   Forecast Bias   Variance  
0            0.8  12.466667  
1            NaN        NaN  
2            NaN        NaN  
3            NaN        NaN  
4            NaN        NaN

Once the metrics are calculated and added to the dataframe, we can analyze the forecast accuracy and demand variability for the respective product. 

We can see that the MAPE value for the Bread forecast is pretty low, which indicates that the forecast for Bread is off by around 1.8%, suggesting that it's an accurate forecast.

BANANA

In [46]:
bananas=pd.DataFrame()
bananas=pd.concat([bananas,df['Date'],df['Bananas_Sales (Pounds)'],df['Bananas_Forecast (Pounds)']],axis=1)
bananas.head()


Date  Bananas_Sales (Pounds)  Bananas_Forecast (Pounds)
0  1/1/2024                     300                        310
1  1/2/2024                     310                        300
2  1/3/2024                     295                        300
3  1/4/2024                     280                        275
4  1/5/2024                     290                        295

In [47]:
bananas_actual=bananas['Bananas_Sales (Pounds)']
bananas_actual.head()

0    300
1    310
2    295
3    280
4    290
Name: Bananas_Sales (Pounds), dtype: int64

In [48]:
bananas_fore=bananas['Bananas_Forecast (Pounds)']
bananas_fore.head()

0    310
1    300
2    300
3    275
4    295
Name: Bananas_Forecast (Pounds), dtype: int64

In [49]:
error=[]
abs_error=[]
perc_error=[]
sq_error=[]
for i in range(len(bananas_actual)):
    ele1=bananas_actual[i]-bananas_fore[i]
    ele2=int(math.fabs(ele1))
    ele3=(ele2/bananas_actual[i])*100
    ele4=ele1**2
    error.append(ele1)
    abs_error.append(ele2)
    perc_error.append(ele3)
    sq_error.append(ele4)

print(error)
print(abs_error)
print(perc_error)
print(sq_error)

[-10, 10, -5, 5, -5, -5, -10, -5, 10, -5, -5, -10, 10, -5, 5, -5, -5, -10, -5, 10, -5, -5, -10, 10, -5, 5, -5, -5, -10, -5]
[10, 10, 5, 5, 5, 5, 10, 5, 10, 5, 5, 10, 10, 5, 5, 5, 5, 10, 5, 10, 5, 5, 10, 10, 5, 5, 5, 5, 10, 5]
[3.3333333333333335, 3.225806451612903, 1.694915254237288, 1.7857142857142856, 1.7241379310344827, 1.7543859649122806, 3.3333333333333335, 1.5625, 3.225806451612903, 1.694915254237288, 1.7857142857142856, 3.3333333333333335, 3.225806451612903, 1.694915254237288, 1.7857142857142856, 1.7241379310344827, 1.7543859649122806, 3.3333333333333335, 1.5625, 3.225806451612903, 1.694915254237288, 1.7857142857142856, 3.3333333333333335, 3.225806451612903, 1.694915254237288, 1.7857142857142856, 1.7241379310344827, 1.7543859649122806, 3.3333333333333335, 1.5625]
[100, 100, 25, 25, 25, 25, 100, 25, 100, 25, 25, 100, 100, 25, 25, 25, 25, 100, 25, 100, 25, 25, 100, 100, 25, 25, 25, 25, 100, 25]


In [50]:
df1=pd.DataFrame(error)
df2=pd.DataFrame(abs_error)
df3=pd.DataFrame(perc_error)
df4=pd.DataFrame(sq_error)


In [51]:
bananas=pd.concat([bananas,df1,df2,df3,df4,],axis=1)

In [52]:
bananas.columns=['Date','Bananas_Sales (Pounds)','Bananas_Forecast (Pounds)','Error','Absolute Error','Percentage Error','Squared Error' ]

In [53]:
bananas.head()

Date  Bananas_Sales (Pounds)  Bananas_Forecast (Pounds)  Error  \
0  1/1/2024                     300                        310    -10   
1  1/2/2024                     310                        300     10   
2  1/3/2024                     295                        300     -5   
3  1/4/2024                     280                        275      5   
4  1/5/2024                     290                        295     -5   

   Absolute Error  Percentage Error  Squared Error  
0              10          3.333333            100  
1              10          3.225806            100  
2               5          1.694915             25  
3               5          1.785714             25  
4               5          1.724138             25

In [54]:
mape_val=[]
for i in range(len(bananas_actual)):
    ele=math.fabs((bananas_actual[i]-bananas_fore[i])/bananas_actual[i])*100
    mape_val.append(ele)
print(mape_val)

[3.3333333333333335, 3.225806451612903, 1.694915254237288, 1.7857142857142856, 1.7241379310344827, 1.7543859649122806, 3.3333333333333335, 1.5625, 3.225806451612903, 1.694915254237288, 1.7857142857142856, 3.3333333333333335, 3.225806451612903, 1.694915254237288, 1.7857142857142856, 1.7241379310344827, 1.7543859649122806, 3.3333333333333335, 1.5625, 3.225806451612903, 1.694915254237288, 1.7857142857142856, 3.3333333333333335, 3.225806451612903, 1.694915254237288, 1.7857142857142856, 1.7241379310344827, 1.7543859649122806, 3.3333333333333335, 1.5625]


In [55]:
Sum=0
for i in mape_val:
    Sum+=i
mape=Sum/len(mape_val)
print(mape)

2.2885083881887556


In [56]:
MAPE=[mape]
df5=pd.DataFrame(MAPE)
bananas=pd.concat([bananas,df5],axis=1)

In [57]:
mad_val=[]
for i in range(len(bananas_actual)):
    ele=math.fabs(bananas_actual[i]-bananas_fore[i])
    mad_val.append(ele)
print(mad_val)

[10.0, 10.0, 5.0, 5.0, 5.0, 5.0, 10.0, 5.0, 10.0, 5.0, 5.0, 10.0, 10.0, 5.0, 5.0, 5.0, 5.0, 10.0, 5.0, 10.0, 5.0, 5.0, 10.0, 10.0, 5.0, 5.0, 5.0, 5.0, 10.0, 5.0]


In [58]:
summ=0
for i in mad_val:
    summ+=i
mad=Sum/len(mad_val)
print(mad)

2.2885083881887556


In [59]:
MAD=[mad]
df6=pd.DataFrame(MAD)
bananas=pd.concat([bananas,df6],axis=1)

In [60]:
fore_val=[]
for i in range(len(bananas_fore)):
    ele=bananas_fore[i]-bananas_actual[i]
    fore_val.append(ele)
print(fore_val)

[10, -10, 5, -5, 5, 5, 10, 5, -10, 5, 5, 10, -10, 5, -5, 5, 5, 10, 5, -10, 5, 5, 10, -10, 5, -5, 5, 5, 10, 5]


In [61]:
ssum=0
for i in fore_val:
    ssum+=i
fore=ssum/len(fore_val)
print(fore)

2.5


In [62]:
BIAS=[fore]
df7=pd.DataFrame(BIAS)
bananas=pd.concat([bananas,df7],axis=1)

In [63]:
sse=bananas["Squared Error"]
suum=0
for i in sse:
    suum+=i
var=suum/len(sse)
print(var)

52.5


In [64]:
VAR=[var]
df8=pd.DataFrame(VAR)
bananas=pd.concat([bananas,df8],axis=1)

In [65]:
bananas.columns=['Date','Bananas_Sales (Pounds)','Bananas_Forecast (Pounds)','Error','Absolute Error','Percentage Error','Squared Error','MAPE','MAD','Forecast Bias','Variance']

In [66]:
bananas.head()

Date  Bananas_Sales (Pounds)  Bananas_Forecast (Pounds)  Error  \
0  1/1/2024                     300                        310    -10   
1  1/2/2024                     310                        300     10   
2  1/3/2024                     295                        300     -5   
3  1/4/2024                     280                        275      5   
4  1/5/2024                     290                        295     -5   

   Absolute Error  Percentage Error  Squared Error      MAPE       MAD  \
0              10          3.333333            100  2.288508  2.288508   
1              10          3.225806            100       NaN       NaN   
2               5          1.694915             25       NaN       NaN   
3               5          1.785714             25       NaN       NaN   
4               5          1.724138             25       NaN       NaN   

   Forecast Bias  Variance  
0            2.5      52.5  
1            NaN       NaN  
2            NaN       NaN  
3            NaN       NaN  
4            NaN       NaN

Once the metrics are calculated and added to the dataframe, we can analyze the forecast accuracy and demand variability for the respective product. 

We can see that the MAPE value for the Bananas forecast is pretty low, which indicates that the forecast for Bananas is off by around 2.3%, suggesting that it's an accurate forecast, upto a certain extent

CHICKEN BREAST

In [88]:
chic=pd.DataFrame()
chic=pd.concat([chic,df['Date'],df['Chicken_Sales (Pounds)'],df['Chicken_Forecast (Pounds)']],axis=1)
chic.head()


Date  Chicken_Sales (Pounds)  Chicken_Forecast (Pounds)
0  1/1/2024                     180                        175
1  1/2/2024                     175                        170
2  1/3/2024                     190                        185
3  1/4/2024                     185                        190
4  1/5/2024                     170                        165

In [89]:
chic_actual=chic['Chicken_Sales (Pounds)']
chic_actual.head()

0    180
1    175
2    190
3    185
4    170
Name: Chicken_Sales (Pounds), dtype: int64

In [90]:
chic_fore=chic['Chicken_Forecast (Pounds)']
chic_fore.head()

0    175
1    170
2    185
3    190
4    165
Name: Chicken_Forecast (Pounds), dtype: int64

In [91]:
error=[]
abs_error=[]
perc_error=[]
sq_error=[]
for i in range(len(chic_actual)):
    ele1=chic_actual[i]-chic_fore[i]
    ele2=int(math.fabs(ele1))
    ele3=(ele2/chic_actual[i])*100
    ele4=ele1**2
    error.append(ele1)
    abs_error.append(ele2)
    perc_error.append(ele3)
    sq_error.append(ele4)

print(error)
print(abs_error)
print(perc_error)
print(sq_error)

[5, 5, 5, -5, 5, 0, -5, -5, 5, 5, -5, 5, 0, -5, -5, 5, 5, -5, 5, 0, -5, -5, 5, 5, -5, 5, 0, -5, -5, 5]
[5, 5, 5, 5, 5, 0, 5, 5, 5, 5, 5, 5, 0, 5, 5, 5, 5, 5, 5, 0, 5, 5, 5, 5, 5, 5, 0, 5, 5, 5]
[2.7777777777777777, 2.857142857142857, 2.631578947368421, 2.7027027027027026, 2.941176470588235, 0.0, 3.3333333333333335, 2.7777777777777777, 2.857142857142857, 2.631578947368421, 2.7027027027027026, 2.941176470588235, 0.0, 3.3333333333333335, 2.7777777777777777, 2.857142857142857, 2.631578947368421, 2.7027027027027026, 2.941176470588235, 0.0, 3.3333333333333335, 2.7777777777777777, 2.857142857142857, 2.631578947368421, 2.7027027027027026, 2.941176470588235, 0.0, 3.3333333333333335, 2.7777777777777777, 2.857142857142857]
[25, 25, 25, 25, 25, 0, 25, 25, 25, 25, 25, 25, 0, 25, 25, 25, 25, 25, 25, 0, 25, 25, 25, 25, 25, 25, 0, 25, 25, 25]


In [92]:
df1=pd.DataFrame(error)
df2=pd.DataFrame(abs_error)
df3=pd.DataFrame(perc_error)
df4=pd.DataFrame(sq_error)


In [93]:
chic=pd.concat([chic,df1,df2,df3,df4,],axis=1)

In [94]:
chic.columns=['Date','Chicken_Sales (Pounds)','Chicken_Forecast (Pounds)','Error','Absolute Error','Percentage Error','Squared Error' ]

In [95]:
chic.head()

Date  Chicken_Sales (Pounds)  Chicken_Forecast (Pounds)  Error  \
0  1/1/2024                     180                        175      5   
1  1/2/2024                     175                        170      5   
2  1/3/2024                     190                        185      5   
3  1/4/2024                     185                        190     -5   
4  1/5/2024                     170                        165      5   

   Absolute Error  Percentage Error  Squared Error  
0               5          2.777778             25  
1               5          2.857143             25  
2               5          2.631579             25  
3               5          2.702703             25  
4               5          2.941176             25

In [96]:
mape_val=[]
for i in range(len(chic_actual)):
    ele=math.fabs((chic_actual[i]-chic_fore[i])/chic_actual[i])*100
    mape_val.append(ele)
print(mape_val)

[2.7777777777777777, 2.857142857142857, 2.631578947368421, 2.7027027027027026, 2.941176470588235, 0.0, 3.3333333333333335, 2.7777777777777777, 2.857142857142857, 2.631578947368421, 2.7027027027027026, 2.941176470588235, 0.0, 3.3333333333333335, 2.7777777777777777, 2.857142857142857, 2.631578947368421, 2.7027027027027026, 2.941176470588235, 0.0, 3.3333333333333335, 2.7777777777777777, 2.857142857142857, 2.631578947368421, 2.7027027027027026, 2.941176470588235, 0.0, 3.3333333333333335, 2.7777777777777777, 2.857142857142857]


In [97]:
Sum=0
for i in mape_val:
    Sum+=i
mape=Sum/len(mape_val)
print(mape)

2.4869922996857974


In [98]:
MAPE=[mape]
df5=pd.DataFrame(MAPE)
chic=pd.concat([chic,df5],axis=1)

In [99]:
mad_val=[]
for i in range(len(chic_actual)):
    ele=math.fabs(chic_actual[i]-chic_fore[i])
    mad_val.append(ele)
print(mad_val)

[5.0, 5.0, 5.0, 5.0, 5.0, 0.0, 5.0, 5.0, 5.0, 5.0, 5.0, 5.0, 0.0, 5.0, 5.0, 5.0, 5.0, 5.0, 5.0, 0.0, 5.0, 5.0, 5.0, 5.0, 5.0, 5.0, 0.0, 5.0, 5.0, 5.0]


In [100]:
summ=0
for i in mad_val:
    summ+=i
mad=Sum/len(mad_val)
print(mad)

2.4869922996857974


In [101]:
MAD=[mad]
df6=pd.DataFrame(MAD)
chic=pd.concat([chic,df6],axis=1)

In [102]:
fore_val=[]
for i in range(len(chic_fore)):
    ele=chic_fore[i]-chic_actual[i]
    fore_val.append(ele)
print(fore_val)

[-5, -5, -5, 5, -5, 0, 5, 5, -5, -5, 5, -5, 0, 5, 5, -5, -5, 5, -5, 0, 5, 5, -5, -5, 5, -5, 0, 5, 5, -5]


In [103]:
ssum=0
for i in fore_val:
    ssum+=i
fore=ssum/len(fore_val)
print(fore)

-0.3333333333333333


In [104]:
BIAS=[fore]
df7=pd.DataFrame(BIAS)
chic=pd.concat([chic,df7],axis=1)

In [105]:
sse=chic["Squared Error"]
suum=0
for i in sse:
    suum+=i
var=suum/len(sse)
print(var)

21.666666666666668


In [106]:
VAR=[var]
df8=pd.DataFrame(VAR)
chic=pd.concat([chic,df8],axis=1)

In [138]:
chic.columns=['Date','Chicken_Sales (Pounds)','Chicken_Forecast (Pounds)','Error','Absolute Error','Percentage Error','Squared Error','MAPE','MAD','Forecast Bias','Variance']

In [108]:
chic.head()

Date  Chicken_Sales (Pounds)  Chicken_Forecast (Pounds)  Error  \
0  1/1/2024                     180                        175      5   
1  1/2/2024                     175                        170      5   
2  1/3/2024                     190                        185      5   
3  1/4/2024                     185                        190     -5   
4  1/5/2024                     170                        165      5   

   Absolute Error  Percentage Error  Squared Error      MAPE       MAD  \
0               5          2.777778             25  2.486992  2.486992   
1               5          2.857143             25       NaN       NaN   
2               5          2.631579             25       NaN       NaN   
3               5          2.702703             25       NaN       NaN   
4               5          2.941176             25       NaN       NaN   

   Forecast Bias   Variance  
0      -0.333333  21.666667  
1            NaN        NaN  
2            NaN        NaN  
3            NaN        NaN  
4            NaN        NaN

Once the metrics are calculated and added to the dataframe, we can analyze the forecast accuracy and demand variability for the respective product. 

We can see that the MAPE value for the Chicken forecast is pretty low, which indicates that the forecast for Chicken is off by around 2.5%, suggesting that it's an accurate forecast, upto a certain extent

TOILET PAPER

In [109]:
tp=pd.DataFrame()
tp=pd.concat([tp,df['Date'],df['Toilet_Paper_Sales (Packs of 12 Rolls)'],df['Toilet_Paper_Forecast (Packs of 12 Rolls)']],axis=1)
tp.head()


Date  Toilet_Paper_Sales (Packs of 12 Rolls)  \
0  1/1/2024                                     100   
1  1/2/2024                                     110   
2  1/3/2024                                     120   
3  1/4/2024                                     130   
4  1/5/2024                                     105   

   Toilet_Paper_Forecast (Packs of 12 Rolls)  
0                                        105  
1                                        115  
2                                        125  
3                                        135  
4                                        100

In [112]:
tp_actual=tp['Toilet_Paper_Sales (Packs of 12 Rolls)']
tp_actual.head()

0    100
1    110
2    120
3    130
4    105
Name: Toilet_Paper_Sales (Packs of 12 Rolls), dtype: int64

In [113]:
tp_fore=tp['Toilet_Paper_Forecast (Packs of 12 Rolls)']
tp_fore.head()

0    105
1    115
2    125
3    135
4    100
Name: Toilet_Paper_Forecast (Packs of 12 Rolls), dtype: int64

In [114]:
error=[]
abs_error=[]
perc_error=[]
sq_error=[]
for i in range(len(tp_actual)):
    ele1=tp_actual[i]-tp_fore[i]
    ele2=int(math.fabs(ele1))
    ele3=(ele2/tp_actual[i])*100
    ele4=ele1**2
    error.append(ele1)
    abs_error.append(ele2)
    perc_error.append(ele3)
    sq_error.append(ele4)

print(error)
print(abs_error)
print(perc_error)
print(sq_error)

[-5, -5, -5, -5, 5, 5, 5, 5, 5, -5, 5, -5, 5, 5, -5, -5, -5, -5, 5, 5, 5, 5, 5, -5, 5, -5, 5, 5, -5, -5]
[5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5]
[5.0, 4.545454545454546, 4.166666666666666, 3.8461538461538463, 4.761904761904762, 4.3478260869565215, 4.0, 3.7037037037037033, 3.571428571428571, 4.545454545454546, 4.3478260869565215, 4.166666666666666, 4.0, 3.8461538461538463, 5.0, 4.545454545454546, 4.166666666666666, 3.8461538461538463, 4.761904761904762, 4.3478260869565215, 4.0, 3.7037037037037033, 3.571428571428571, 4.545454545454546, 4.3478260869565215, 4.166666666666666, 4.0, 3.8461538461538463, 5.0, 4.545454545454546]
[25, 25, 25, 25, 25, 25, 25, 25, 25, 25, 25, 25, 25, 25, 25, 25, 25, 25, 25, 25, 25, 25, 25, 25, 25, 25, 25, 25, 25, 25]


In [115]:
df1=pd.DataFrame(error)
df2=pd.DataFrame(abs_error)
df3=pd.DataFrame(perc_error)
df4=pd.DataFrame(sq_error)


In [116]:
tp=pd.concat([tp,df1,df2,df3,df4,],axis=1)

In [117]:
tp.columns=['Date','Toiler_Paper_Sales (Packs of 12 Rolls)','Toilet_Paper_Forecast (Packs of 12 Rolls)','Error','Absolute Error','Percentage Error','Squared Error' ]

In [118]:
tp.head()

Date  Toiler_Paper_Sales (Packs of 12 Rolls)  \
0  1/1/2024                                     100   
1  1/2/2024                                     110   
2  1/3/2024                                     120   
3  1/4/2024                                     130   
4  1/5/2024                                     105   

   Toilet_Paper_Forecast (Packs of 12 Rolls)  Error  Absolute Error  \
0                                        105     -5               5   
1                                        115     -5               5   
2                                        125     -5               5   
3                                        135     -5               5   
4                                        100      5               5   

   Percentage Error  Squared Error  
0          5.000000             25  
1          4.545455             25  
2          4.166667             25  
3          3.846154             25  
4          4.761905             25

In [119]:
mape_val=[]
for i in range(len(tp_actual)):
    ele=math.fabs((tp_actual[i]-tp_fore[i])/tp_actual[i])*100
    mape_val.append(ele)
print(mape_val)

[5.0, 4.545454545454546, 4.166666666666666, 3.8461538461538463, 4.761904761904762, 4.3478260869565215, 4.0, 3.7037037037037033, 3.571428571428571, 4.545454545454546, 4.3478260869565215, 4.166666666666666, 4.0, 3.8461538461538463, 5.0, 4.545454545454546, 4.166666666666666, 3.8461538461538463, 4.761904761904762, 4.3478260869565215, 4.0, 3.7037037037037033, 3.571428571428571, 4.545454545454546, 4.3478260869565215, 4.166666666666666, 4.0, 3.8461538461538463, 5.0, 4.545454545454546]


In [120]:
Sum=0
for i in mape_val:
    Sum+=i
mape=Sum/len(mape_val)
print(mape)

4.2414644400151635


In [124]:
MAPE=[mape]
df5=pd.DataFrame(MAPE)
tp=pd.concat([tp,df5],axis=1)

In [122]:
mad_val=[]
for i in range(len(tp_actual)):
    ele=math.fabs(tp_actual[i]-tp_fore[i])
    mad_val.append(ele)
print(mad_val)

[5.0, 5.0, 5.0, 5.0, 5.0, 5.0, 5.0, 5.0, 5.0, 5.0, 5.0, 5.0, 5.0, 5.0, 5.0, 5.0, 5.0, 5.0, 5.0, 5.0, 5.0, 5.0, 5.0, 5.0, 5.0, 5.0, 5.0, 5.0, 5.0, 5.0]


In [123]:
summ=0
for i in mad_val:
    summ+=i
mad=Sum/len(mad_val)
print(mad)

4.2414644400151635


In [129]:
MAD=[mad]
df6=pd.DataFrame(MAD)
tp=pd.concat([tp,df6],axis=1)

In [130]:
fore_val=[]
for i in range(len(tp_fore)):
    ele=tp_fore[i]-tp_actual[i]
    fore_val.append(ele)
print(fore_val)

[5, 5, 5, 5, -5, -5, -5, -5, -5, 5, -5, 5, -5, -5, 5, 5, 5, 5, -5, -5, -5, -5, -5, 5, -5, 5, -5, -5, 5, 5]


In [131]:
ssum=0
for i in fore_val:
    ssum+=i
fore=ssum/len(fore_val)
print(fore)

-0.3333333333333333


In [133]:
BIAS=[fore]
df7=pd.DataFrame(BIAS)
tp=pd.concat([tp,df7],axis=1)

In [135]:
sse=tp["Squared Error"]
suum=0
for i in sse:
    suum+=i
var=suum/len(sse)
print(var)

25.0


In [136]:
VAR=[var]
df8=pd.DataFrame(VAR)
tp=pd.concat([tp,df8],axis=1)

In [139]:
tp.columns=['Date','Toilet_Paper_Sales (Packs of 12 Rolls)','Toilet_Paper_Forecast (Packs of 12 Rolls)','Error','Absolute Error','Percentage Error','Squared Error','MAPE','MAD','Forecast Bias','Variance']

In [141]:
tp.head()

Date  Toilet_Paper_Sales (Packs of 12 Rolls)  \
0  1/1/2024                                     100   
1  1/2/2024                                     110   
2  1/3/2024                                     120   
3  1/4/2024                                     130   
4  1/5/2024                                     105   

   Toilet_Paper_Forecast (Packs of 12 Rolls)  Error  Absolute Error  \
0                                        105     -5               5   
1                                        115     -5               5   
2                                        125     -5               5   
3                                        135     -5               5   
4                                        100      5               5   

   Percentage Error  Squared Error      MAPE       MAD  Forecast Bias  \
0          5.000000             25  4.241464  4.241464      -0.333333   
1          4.545455             25       NaN       NaN            NaN   
2          4.166667             25       NaN       NaN            NaN   
3          3.846154             25       NaN       NaN            NaN   
4          4.761905             25       NaN       NaN            NaN   

   Variance  
0      25.0  
1       NaN  
2       NaN  
3       NaN  
4       NaN

Once the metrics are calculated and added to the dataframe, we can analyze the forecast accuracy and demand variability for the respective product. 

We can see that the MAPE value for the Toilet Paper forecast is low, which indicates that the forecast for Toilet Paper is off by around 4.2%, suggesting that it's somewhat accurate, but it could be better, and is inaccurate compared to the other four products available.

Overall, we can say that the forecast accuracy of all the products together are very accurate, which will lead to good customer satisfaction and inventory management.